src/fdc/DiskImageUtils.cc

```diff
@@ -170,32 +205,49 @@ static SetBootSectorResult setBootSector(
 		nbSectors = 720;	// force nbSectors to 720, why?
 	}
 
-	boot.nrSectors = uint16_t(nbSectors); // TODO check for overflow
+	boot.nrSectors = nbSectors < 0x10000 ? narrow<uint16_t>(nbSectors) : 0;
+	if (bootType != MSXBootSectorType::EXTENDED && boot.nrSectors == 0) {
+		throw CommandException("Too many sectors for FAT12 ", nbSectors);
+	}
+
 	boot.nrSides = nbSides;
 	boot.spCluster = nbSectorsPerCluster;
 	boot.nrFats = nbFats;
 	boot.sectorsFat = nbSectorsPerFat;
 	boot.dirEntries = nbDirEntry;
 	boot.descriptor = descriptor;
 	boot.resvSectors = nbReservedSectors;
-	boot.hiddenSectors = nbHiddenSectors;
 
-	if (!dos1) {
-		// set random volume id
-		boot.vol_id = random_32bit() & 0x7F7F7F7F; // why are bits masked?
+	if (bootType == MSXBootSectorType::DOS1) {
+		auto params = boot.params.dos1;
```

Comment:
This seems like a bug. It should be:
`auto& params = boot.params.dos1;`
You want to refer to part of the boot sector and make changes there. Instead of copying part in a local variable and changing that copy (and then immediately discarding that changed copy).

Same on line 225 and 229.

Reply:
Oops. My testing didn’t catch it because I based the boot block template on a 64 MB image. Let me zero out the sector count in the boot block template to make sure it fails if incorrect.